### PR TITLE
Add unit test to ensure no gradients sync when calling ddp.module(input)

### PR DESCRIPTION
**Pull request by chenyangyu1988**
Summary:
Add unit test to ensure no gradients sync when calling ddp.module(input), e.g not invoking prepare_for_backward

PyText is depending on DDP for data parallel distributed training. To support accumulate gradients locally before gradients sync, we are calling orig_model.forward instead of ddp_model.forward. Add a unit test to avoid changes break the assumption.

Differential Revision: D15263155

### CODE REVIEW

File: test/test_c10d.py [hunk 1]
```diff
@@ -2370,6 +2370,57 @@ def check_no_grads():
...
+            loss = F.mse_loss(output, target.to(output.device))
+            loss.backward()
+
+        # ensure accumulate grads works with no_grad
+        with torch.no_grad():
+            ddp_model.train()
+            ddp_model.module(input)
+
+        # check two model parameters over 2 iterations
+        for iteration in range(2):
```
Comment (mrshenli) [inline test/test_c10d.py]: Let's do 4 iterations to make sure that things are still correct after the first sync.

File: test/test_c10d.py [hunk 2]
```diff
@@ -2370,6 +2370,57 @@ def check_no_grads():
         # No parameter should have their gradient set.
         check_no_grads()
 
+    @skip_if_not_multigpu
+    @skip_if_not_nccl
+    def test_accumulate_gradients(self):
```
Comment (mrshenli) [inline test/test_c10d.py]: Shall we add a comment to say that this would only work when this is a single model replica per process.

**Review by mrshenli — APPROVED** [review]
Thanks for adding the test. I feel this test should only temporarily live to make sure changes in PyTorch does not break your use cases. And it should retire when we have a better solution (e.g., a context manager as @apaszke suggested).
